### PR TITLE
feat(mcp): renew the stored SSO identity assertion behind ID-JAG

### PR DESCRIPTION
**Pull request by yassin-berriai**
## TLDR

Problem this solves:

- ID-JAG agents stop working an id_token lifetime after login
- Only recovery today is another interactive SSO sign-in
- The stored refresh token was never redeemed

How it solves it:

- Renew the stored assertion lazily when it nears expiry
- One redemption per user, single-flighted across replicas
- Rotation written back; a refusal still challenges the user

## Linear ticket

Resolves LIT-5059

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Captured at commit `1652defe6c` against a live proxy on localhost:4459 talking to a local OIDC provider that also plays the ID-JAG org AS, the upstream resource AS and the upstream MCP server, every leg verifying signatures and claims. Its id_tokens live 120 seconds so the expiry a real agent hits after an hour happens inside a two minute capture, and it accepts only HTTP Basic client authentication, the way an IdP application registered for `client_secret_basic` behaves. No mocks anywhere in the path: the proxy signs into that provider for real, redeems the refresh token against its real token endpoint, and the upstream refuses any bearer that did not come out of leg 2. The code has moved on by one change since that transcript was taken, which selects the client-authentication method from the flag the login reads rather than assuming one; the deployment captured here leaves that flag unset and so resolves to the same method it used during the run, and both branches of that selection are covered by tests rather than by this capture.

Before (this is today's behavior, reproduced on this build by leaving `offline_access` out of `GENERIC_SCOPE`, so no refresh token is ever captured and there is nothing to renew):

```
$ curl -s -L -c c.txt -o /dev/null -w 'login: %{url_effective}\n' http://127.0.0.1:4459/sso/key/generate
login: http://127.0.0.1:4459/ui/?login=success

$ curl -s http://127.0.0.1:9459/grants   # what the IdP was asked for
  {'grant': 'authorize', 'scope': 'openid email profile'}
  {'grant': 'authorization_code', 'offline_access': False}

# right after the sign-in
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" \
    -H 'Content-Type: application/json' -H 'Accept: application/json, text/event-stream' \
    -d '{"jsonrpc":"2.0","id":1,"method":"tools/call","params":{"name":"idjag_upstream-whoami","arguments":{}}}'
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"the upstream served agent-owner@example.com"}],"isError":false}}

# 128s later, past the id_token's exp
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" ... (same body)
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"Error: precondition required: The stored IdP identity assertion for this user has expired. Sign in through LiteLLM SSO again to capture a current one."}],"isError":true}}

$ grep 'ID-JAG:' litellm.log | tail -1
ID-JAG: the stored IdP identity assertion for user_id=agent-owner@example.com has expired and no refresh
token was captured with it, so it cannot be renewed without another interactive sign-in. Add
'offline_access' to GENERIC_SCOPE so the SSO login captures one.
```

After (same proxy, same upstream, `GENERIC_SCOPE="openid email profile offline_access"`):

```
$ curl -s -L -c c.txt -o /dev/null -w 'login: %{url_effective}\n' http://127.0.0.1:4459/sso/key/generate
login: http://127.0.0.1:4459/ui/?login=success

$ curl -s http://127.0.0.1:9459/grants
  {'grant': 'authorize', 'scope': 'openid email profile offline_access'}
  {'grant': 'authorization_code', 'offline_access': True}

# an agent gets a LiteLLM key for that user; it never sees an IdP credential
$ AGENT_KEY=$(curl -s -X POST http://127.0.0.1:4459/key/generate -H "Authorization: Bearer $MASTER" \
    -H 'Content-Type: application/json' \
    -d '{"user_id":"agent-owner@example.com","object_permission":{"mcp_servers":["95ad6b0f9ae86407d9f8244bd2f2306a"]}}' \
    | python3 -c 'import sys,json;print(json.load(sys.stdin)["key"])')

# right after the sign-in
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" \
    -H 'Content-Type: application/json' -H 'Accept: application/json, text/event-stream' \
    -d '{"jsonrpc":"2.0","id":1,"method":"tools/call","params":{"name":"idjag_upstream-whoami","arguments":{}}}'
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"the upstream served agent-owner@example.com"}],"isError":false}}

$ curl -s http://127.0.0.1:9459/grants
  {'grant': 'id-jag',     'result': 'ok', 'subject_jti': '1ff403bae3eb423fb62880cf3085962f'}
  {'grant': 'jwt-bearer', 'result': 'ok', 'sub': 'agent-owner'}

# 125s later, past the id_token's exp; this is the call that used to be a 412
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" ... (same body)
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"the upstream served agent-owner@example.com"}],"isError":false}}

$ curl -s http://127.0.0.1:9459/grants   # exactly one refresh, and leg 1 now asserts a different token
  {'grant': 'refresh_token', 'result': 'ok', 'generation': 2}
  {'grant': 'id-jag',        'result': 'ok', 'subject_jti': '23c85e108f18437fa3ac6b6937c2177c'}
  {'grant': 'jwt-bearer',    'result': 'ok', 'sub': 'agent-owner'}

# the rotation was written back
$ docker exec pg psql -U litellm -d litellm -t -A -c 'select "user_id", "updated_at" from "LiteLLM_SSOIdentityAssertion";'
before: agent-owner@example.com|2026-07-31 20:14:49.423
after:  agent-owner@example.com|2026-07-31 20:16:55.316
```

And the settled-failure path, so a dead refresh token still tells the user what to do rather than looping:

```
$ curl -s -X POST http://127.0.0.1:9459/revoke    # the IdP kills the refresh token
{"revoked":true}

# once the renewed id_token also runs out
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" ... (same body)
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"Error: precondition required: The stored IdP identity assertion for this user has expired. Sign in through LiteLLM SSO again to capture a current one."}],"isError":true}}

$ curl -s http://127.0.0.1:9459/grants
  {'grant': 'refresh_token', 'result': 'invalid_grant'}
```

One redemption attempt, then the same challenge as before this PR.

The renewal authenticates the way the login did. Against an IdP application registered for `client_secret_basic`, posting the credentials in the form body instead fails outright, which the same rig shows directly:

```
# with the credentials in the form body
$ curl -s -X POST http://127.0.0.1:4459/mcp/ -H "x-litellm-api-key: Bearer $AGENT_KEY" ... (past expiry)
data: {"jsonrpc":"2.0","id":1,"result":{"content":[{"type":"text","text":"Error: precondition required: The stored IdP identity assertion for this user has expired. Sign in through LiteLLM SSO again to capture a current one."}],"isError":true}}

$ curl -s http://127.0.0.1:9459/grants
  {'grant': 'refresh_token', 'result': 'no_basic_auth'}
```

Worth stating because it is the reason that counterfactual exists at all: the first version of this rig accepted both client-authentication methods, and a green capture against it said nothing about which method the renewal sent. A test double more permissive than production hides precisely the defects that matter, so the rig was narrowed to refuse everything the real IdP application would refuse before any of this was believed.

## Type

🆕 New Feature

## Changes

`RefreshingSSOAssertionStore` wraps the database reader and satisfies the same `SSOAssertionStore` protocol, so the `id_jag` arm is untouched: it still reads one assertion and still judges expiry itself. A read that finds an assertion within 60 seconds of its `exp` runs one renewal and then every caller re-reads the store, which stays the authority; an assertion comfortably inside its lifetime costs exactly what it did before.

Client authentication goes through the package's shared `build_token_endpoint_client_auth`, as every other token-endpoint POST in the MCP gateway does, and picks the method the login itself used. That is not a constant: the non-PKCE path always sends HTTP Basic, while the PKCE path puts the credentials in the body when `GENERIC_INCLUDE_CLIENT_ID` is set, and both capture assertions. An IdP application may be registered for only one of the two, so the renewal reads the same flag the login reads rather than assuming; picking either one unconditionally 401s half the deployments and produces the sign-in-again 412 this PR exists to remove.

The refresh is redeemed against the generic-OIDC client the login itself used, `GENERIC_TOKEN_ENDPOINT` with `GENERIC_CLIENT_ID` and `GENERIC_CLIENT_SECRET`, which the proxy already materializes from the stored SSO row into the process environment at startup. An assertion can only exist if that client minted it, since the assertion is captured on the generic-OIDC callback and nowhere else, so no new configuration is introduced.

Renewal is single-flighted per user through the same `RefreshCoordinator` the `authorization_code` arm uses, cross-replica when Redis is wired and in-process otherwise. That matters because an IdP that rotates refresh tokens reads two concurrent redemptions of one token as replay and can revoke the whole grant chain, which would cost the user the sign-in this change exists to avoid. The Redis coordinator construction that was inline in `per_user_oauth_store` moved to `runtime_refresh_coordinator` so both arms elect a winner the same way; that move is behavior-preserving and the `authorization_code` tests pass unchanged.

Failures keep the two outcomes distinguishable. A settled refusal, meaning no refresh token was captured, the SSO client is not configured, the IdP returned a 4xx, or the response carried no id_token, leaves the expired assertion in place so the reader's own guard raises the same sign-in-again 412 as before. A transient failure raises `AssertionStoreUnavailable`, the protocol's existing signal, so the caller gets a 503 instead of being blamed for someone else's outage. That covers an unreachable or 5xx token endpoint and also a renewal the database refused to record: the store is what every caller reads afterwards, so a renewal nobody can see is not a success, and reporting it as settled would tell the user to sign in again over a write failure. No new error shape.

A rotated refresh token replaces the stored one and an omitted one carries forward. The write-back skips when the row moved on while the renewal was in flight, which happens when an interactive sign-in lands mid-renewal; overwriting it would put back a refresh token the IdP has already rotated away.

### Cross-replica losers answer retryable instead of guessing

Under Redis-coordinated renewal, a pod that lost the lock waits for the holder and re-reads the row. A row that is still expiring means the holder's renewal was either refused by the IdP or could not be recorded by the database, and those two leave an identical row. The loser does not redeem the token itself, since the holder may already have rotated it at the IdP and a second redemption of a consumed token looks like a refusal. It answers 503 so the client retries; that retry runs uncontended and settles the outcome on its own, a refusal challenging and a success persisting. On the refusal path the loser pays one retry before the challenge. If the holder rotated the token and the write then failed, that rotation is lost and the next read's refusal challenges, the only honest answer since the rotated token was never recorded. Every read taken inside the coordinator, and the final one after it, bypasses the per-process assertion cache and goes to the database, since a loser holding a cached copy of the expiring row would otherwise keep answering 503 for a cache TTL and, if it later won the lock, redeem a token another replica already rotated. The elected callback also re-reads the row before redeeming, so a rotation another replica landed between the read and the lock is never redeemed a second time. No new distributed state

Worth knowing before this helps anyone: a deployment only captures a refresh token when its `GENERIC_SCOPE` includes `offline_access`, and the default is `openid email profile`. The default is deliberately unchanged, since requesting a scope an IdP app has not granted can fail the authorization request outright and break login for existing deployments. The miss is made discoverable instead: the operator log names both the setting and the value, as shown in the before capture above.

## Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

Link to Devin session: https://app.devin.ai/sessions/651e6ce5ac9f4ded94d9023a15866e92
Open in Devin Desktop: https://app.devin.ai/desktop/session/651e6ce5ac9f4ded94d9023a15866e92?variant=devin
Requested by: @yassin-berriai

### CODE REVIEW

File: litellm/proxy/_experimental/mcp_server/outbound_credentials/sso_assertion_refresher.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        assertion = await self._inner.fetch(user_id)
+        if assertion is None or not assertion_expired(assertion, self._clock() + self._skew):
+            return assertion
+        await self._coordinator().run(
+            user_id,
+            _SINGLE_FLIGHT_KEY,
+            refresh=lambda: self._renew(user_id, assertion),
+            reread=_nothing_to_reread,
+        )
+        return await self._inner.fetch(user_id)
```
[the remark this answers is not in the document]
Comment (yassin-berriai) [reply litellm/proxy/_experimental/mcp_server/outbound_credentials/sso_assertion_refresher.py]: Good catch. A write nobody can read is not a success, so it now returns unavailable and answers 503 instead of 412.

**Comment by yassin-berriai** [comment]
@greptileai please re-review. The P1 is fixed at 1652defe6c: a persist failure now returns a transient failure instead of being swallowed, so the reader answers 503 rather than a sign-in-again 412, and the swallowed-write path no longer discards a renewal the store never recorded.

**Comment by yassin-berriai** [comment]
@greptileai please re-review. The client-authentication method for the refresh now follows the same GENERIC_INCLUDE_CLIENT_ID flag the SSO login branches on, instead of assuming HTTP Basic; the PKCE login path sends credentials in the body when that flag is set, so a constant would have 401'd those deployments.

**Comment by yassin-berriai** [comment]
@greptileai

**Review by yucheng-berri — APPROVED** [review]
